scope.go

```diff
@@ -0,0 +1,604 @@
+package rcmgr
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/libp2p/go-buffer-pool"
+	"github.com/libp2p/go-libp2p-core/network"
+)
+
+// Basic resource mamagement.
+type Resources struct {
+	limit Limit
+
+	nconnsIn, nconnsOut     int
+	nstreamsIn, nstreamsOut int
+	nfd                     int
+
+	memory  int64
+	buffers map[interface{}][]byte
+}
+
+// DAG ResourceScopes.
+// Resources accounts for the node usage, constraints signify
+// the dependencies that constrain resource usage.
+type ResourceScope struct {
+	sync.Mutex
+	done   bool
+	refCnt int
+
+	rc          *Resources
+	constraints []*ResourceScope
+}
+
+var _ network.ResourceScope = (*ResourceScope)(nil)
+var _ network.TransactionalScope = (*ResourceScope)(nil)
+
+func NewResources(limit Limit) *Resources {
+	return &Resources{
+		limit: limit,
+	}
+}
+
+func NewResourceScope(limit Limit, constraints []*ResourceScope) *ResourceScope {
+	return &ResourceScope{
+		rc:          NewResources(limit),
+		constraints: constraints,
+	}
+}
+
+// Resources implementation
+func (rc *Resources) checkMemory(rsvp int64) error {
+	// overflow check; this also has the side-effect that we cannot reserve negative memory.
+	newmem := rc.memory + int64(rsvp)
+	if newmem < rc.memory {
+		return fmt.Errorf("memory reservation overflow: %w", ErrResourceLimitExceeded)
+	}
+
+	// limit check
+	if newmem > rc.limit.GetMemoryLimit() {
+		return fmt.Errorf("cannot reserve memory: %w", ErrResourceLimitExceeded)
+	}
+
+	return nil
+}
+
+func (rc *Resources) releaseBuffers() {
+	for key, buf := range rc.buffers {
+		pool.Put(buf)
+		delete(rc.buffers, key)
+	}
+}
+
+func (rc *Resources) reserveMemory(size int64) error {
+	if err := rc.checkMemory(size); err != nil {
+		return err
+	}
+
+	rc.memory += int64(size)
+	return nil
+}
+
+func (rc *Resources) releaseMemory(size int64) {
+	rc.memory -= size
+
+	// sanity check for bugs upstream
+	if rc.memory < 0 {
+		panic("BUG: too much memory released")
+	}
+}
+
+func (rc *Resources) getBuffer(size int) ([]byte, error) {
+	if err := rc.checkMemory(int64(size)); err != nil {
+		return nil, err
+	}
+
+	buf := pool.Get(size)
+
+	rc.memory += int64(size)
+	if rc.buffers == nil {
+		rc.buffers = make(map[interface{}][]byte)
+	}
+	rc.buffers[buf] = buf
+
+	return buf, nil
+}
+
+func (rc *Resources) growBuffer(oldbuf []byte, newsize int) ([]byte, error) {
+	grow := newsize - len(oldbuf)
+	if err := rc.checkMemory(int64(grow)); err != nil {
+		return nil, err
+	}
+
+	_, ok := rc.buffers[oldbuf]
+	if !ok {
+		return nil, fmt.Errorf("cannot grow unknown buffer")
```

Comment:
This error could be more descriptive. It's invalid to pass a buffer not allocated via this resource struct here.

Reply:
sure will fix.

Reply:
fixed.